cocos/core/memop/pool.ts

```diff
@@ -107,15 +115,25 @@ export class Pool<T> extends ScalableContainer {
      * @zh 尝试缩容对象池，以释放内存。
      */
     public tryShrink (): void {
-        if (this._nextAvail >> 1 > this._elementsPerBatch) {
-            if (this._dtor) {
-                for (let i = this._nextAvail >> 1; i <= this._nextAvail; i++) {
-                    this._dtor(this._freePool[i]);
-                }
+        const freeObjectNumber = this._freePool.length;
+        if (freeObjectNumber <= this._shrinkThreshold) {
+            return;
+        }
+
+        let objectNumberToShrink = 0;
+        if (freeObjectNumber >> 1 > this._shrinkThreshold) {
+            objectNumberToShrink = freeObjectNumber >> 1;
+        } else {
+            objectNumberToShrink = Math.floor((freeObjectNumber - this._shrinkThreshold + 1) / 2);
+        }
+
+        if (this._dtor) {
+            for (let i = this._nextAvail, count = objectNumberToShrink; count > 0; --count, --i) {
```

Comment:
```suggestion
            for (let i = this._nextAvail - objectNumberToShrink; i < this._nextAvail; ++i) {
```

Reply:
Yep, good idea. but should remove until this._nextAvail, so should change it to
```typescript
for (let i = this._nextAvail - objectNumberToShrink + 1;  i <=  this._nextAvail; ++i)
```

Reply:
fixed